### DISCUSSION TITLE
The tag plugin I created to display a list of articles for a specific category only shows two items

### DISCUSSION THREAD

**Discussion by nagongze** [hexojs/hexo#5194] [Question / Answer]
```js
hexo.extend.tag.register('post_list', function(args) {
    var categoryNames = args[0].split(',').map(function(cat) { return cat.trim().toUpperCase(); });
    var result = '<ol>';

    hexo.locals.get('posts').sort('-date').forEach(function(post) {
        if (post.categories.some(function(cat) { return categoryNames.includes(cat.name.trim().toUpperCase()); })) {
            result += '<li><a href="' + post.permalink + '">' + post.title + '</a></li>';
        }
    });

    result += '</ol>';
    return result;
});
```
I don't even know what the reason is. Whether it's the tag syntax or the Front matter, I've copied them, and theoretically, it should have more items, but it just displays two items.

![image](https://user-images.githubusercontent.com/68653451/234049360-a8eeb506-0926-438c-9f1f-433dfa2225fe.png)
![image](https://user-images.githubusercontent.com/68653451/234049518-47d796e7-ee20-4e6f-a869-e086670a6ba8.png)

**Comment by lorezyra** [comment]
Have you run this through VScode debugger?

## Reference
- https://blog.richiebartlett.com/Work_%E4%BB%95%E4%BA%8B/Programming_%E9%96%8B%E7%99%BA/Debugging-HEXO-in-VScode/

**Reply by nagongze** [reply]
I’m sorry, I’m not very good at operating it, and it doesn’t show any related errors.
When I use console.log, the information is insufficient.
But I can’t see where the problem is.

**Reply by lorezyra** [reply]
`console.log` is great when you know what to look for... but it's like walking in the dark without a flashlight. That's why I use VScode debugger to see the data in memory. Then I can make faster corrections to my code. 

Another thought occurred to me... in the rendered HTML, are there only two categories? Or, is it hidden by a CSS rule?

**Reply by nagongze** [reply]
> `console.log` is great when you know what to look for... but it's like walking in the dark without a flashlight. That's why I use VScode debugger to see the data in memory. Then I can make faster corrections to my code.
> 
> Another thought occurred to me... in the rendered HTML, are there only two categories? Or, is it hidden by a CSS rule?

I only have one category, and I used Ctrl+F to search that the names are all the same, and there is no hidden CSS.

**Reply by lorezyra** [reply]
We may be getting lost in translation. I mean to suggest that you review the HTML output to see if all the categories are there. I've made UI's where some data was hidden by a `overflow` or `max-height` CSS rule.

**Comment by yoshinorin** [comment]
@nagongze 

I'm not sure what you want to do, but what about this code?

```js
hexo.extend.tag.register('post_list', (args) => {
  const categoryNames = Array.from(new Set(args[0].split(',').map((cat) => { 
    return cat.trim().toUpperCase();
  })));
  let result = '<ol>';

  hexo.locals.get('posts').sort('-date').forEach((post) => {
    if (post.categories && post.categories.length > 0) {
      const postCategories = post.categories.map((c) => c.name.trim().toUpperCase());
      const matchedCategories = postCategories.filter(pc => {
        return categoryNames.find(cn => cn === pc);
      });
      if (matchedCategories.length > 0) {
        result += '<li><a href="' + post.permalink + '">' + post.title + '</a></li>';
      }
    }
  });

  result += '</ol>';
  return result;
});
```

**Reply by nagongze** [reply]
The result is still two items regardless of whether I check the category command or use folder auto-generation category.
Unless I manually modify the category of an article once in the hexo server status, it will not appear, which means that the online deployment status will only be incorrect.
I want to link articles in the same category like documents do, but I don’t want to manually add post_link every time I write a new article.
That’s what it’s for. However, no matter how I modify the code, the final result will not be all the articles in my category.
But thank you for helping me fix the code.

**Comment by stevenjoezhang** [comment]
@nagongze Your code actually works fine in my blog. Can you provide your source repository, so I can help debugging?

**Comment by jerryc127** [comment]
I hava same problem,  hexo.locals.get('posts') return 2 items rather than all post
https://github.com/orgs/hexojs/discussions/5230

@nagongze how to solve it ?
@stevenjoezhang it is bugs?

**Comment by uiolee** [comment]
I tried it and it works.

test in v6.3 and v7

![图片](https://github.com/hexojs/hexo/assets/22849383/d29d01ec-c73f-486c-89ad-48cc95b5a02f)


![图片](https://github.com/hexojs/hexo/assets/22849383/19e6b0c7-ecf5-4b57-b0ca-de30743c34df)


```js
hexo.extend.tag.register("post_list", function (args) {
  var result = "<ol>";
  console.log(hexo.locals.get('posts').length);
  hexo.locals.get("posts").sort("-date").forEach(function (post) {
      console.log(post.title, post.permalink);
      result += '<li><a href="' + post.permalink + '">' + post.title + "</a></li>";
    });

  result += "</ol>";
  return result;
});
```

**Comment by D-Sketon** [comment]
It's so strange, I changed the number of articles containing the tag and the output is different
one post with tag: 
![image](https://github.com/hexojs/hexo/assets/49871906/19172e80-cfcb-43d7-97a2-dea8c5d1d7cd)
two posts with tag: 
![image](https://github.com/hexojs/hexo/assets/49871906/ea71a76e-6962-4650-9fa1-a91d8358ce24)

**Reply by D-Sketon** [reply]
It seems that with the tag added the output becomes a random event

**Comment by lorezyra** [comment]
Can you output your node environment settings? I'm wondering if you have a poorly configured heap config...
